Refactor refreshes in Host#scan_from_queue

blomquisg opened the pull request:
Having an OpenStack-specific check in the list of refreshes that happen during `scan_from_queue` was really bothering me.

Refactoring to allow `Host` subclasses to append their own refreshes during `scan_from_queue`.

(This will impact #3130 ... so if that gets merged first, I can rebase this one).

/cc @Fryguy, @Ladas

Ladas commented:
nice refactoring :+1:

matthewd commented on app/models/host.rb:
```diff
@@ -1568,6 +1524,30 @@ def scan_from_queue(taskid=nil)
+    # Firewall Rules and Advanced Settings go through EMS so we don't need Host credentials
+    scan_refresh("Firewall Rules", log_target, task) { refresh_firewall_rules }
+    scan_refresh("Advanced Settings", log_target, task) { refresh_advanced_settings }
+    scan_refresh("IPMI information", log_target, task) { refresh_ipmi } unless ext_management_system
```
I think the conditional's getting lost here; consider using a block `unless` instead?

blomquisg replied:
Yeah, it was a last minute decision ... will change back to the block (as it was before)

matthewd commented on app/models/host.rb:
```diff
@@ -1568,6 +1524,30 @@ def scan_from_queue(taskid=nil)
+    task.update_status("Active", "Ok", "Refreshing #{title}") if task
+
+    benchmark_name = "refresh_#{title.parameterize.underscore}".to_sym
+    Benchmark.realtime_block(benchmark_name) { ssu ? yield(ssu) : yield }
```
No need for an `ssu` parameter in this method, AFAICS

blomquisg replied:
Well, it yields two forms ... one with the `ssu` param (hate that name, but kept it for consistency with the things it's yielding to) and one without the `ssu` param.

I really wanted a better way to do this, but couldn't see one that didn't result in ridiculous repeated code (that I was trying to get away from in the first place).

however, if you mean I don't have to wrap the `ssu` in parens, as in:

```ruby
  { ssu ? yield ssu : yield }
```

I actually tried that first and got a syntax error around `ssu` since it couldn't parse what was going on there without the parens.

matthewd commented on app/models/host.rb:
```diff
@@ -1568,6 +1524,30 @@ def scan_from_queue(taskid=nil)
+  end
+
+  def scan_refresh(title, log_target, task, ssu = nil)
+    _log.info("Refreshing #{title} for #{log_target}")
```
I think we can reconstruct `log_target` on demand, instead of passing it in

blomquisg replied:
agreed!

Fryguy commented on app/models/host.rb:
```diff
@@ -1568,6 +1524,33 @@ def scan_from_queue(taskid=nil)
+    end
+  end
+
+  def scan_refreshes_with_connection(task, ssu)
```
Love this PR, but not sure on the method name...connection is so generic that it doesn't explain what kind of connection, which to me makes the code unclear.  Maybe it's not important for others, but I think `scan_refreshes_over_ssh` or something like that might be clearer.

Also, should all these breakout methods be private?

blomquisg replied:
> not sure on the method name ... I think `scan_refreshes_over_ssh` or something like that might be clearer

I like `scan_refreshes_over_ssh`.  Going with that.

> should all these breakout methods be private?

I can make `scan_refresh` private, but I can't make `scan_refreshes` or `scan_refreshes_over_ssh` private, since subclasses should be allowed to append extra logic to them...  I can make them `protected` at best.

matthewd replied:
> I can't make scan_refreshes or scan_refreshes_over_ssh private, since subclasses [..]

FYI, yes you can... ruby private semantics allow that; protected means something else.

matthewd commented on app/models/manageiq/providers/openstack/infra_manager/host.rb:
```diff
@@ -107,4 +107,10 @@ def refresh_openstack_services(ssu)
+
+  def scan_refreshes_with_connection(log_target, task, ssu)
+    super
+    scan_refresh("OpenStack Services", log_target, task, ssu) { |ssu| refresh_openstack_services(ssu) }
```
This method has fallen behind

blomquisg commented:
@matthewd [said](https://github.com/ManageIQ/manageiq/pull/3870#discussion_r37095501):
> This method has fallen behind

Thanks!  Fixed.  Guessing that means there's no test around this :/  Makes refactoring fun :)

Also, addressed @Fryguy's [comments](https://github.com/ManageIQ/manageiq/pull/3870#discussion_r37086291):

chessbyte commented:
@blomquisg What is the status of this PR?

chessbyte commented:
@blomquisg What is the status of this PR?

chessbyte commented:
@blomquisg What is the status of this PR?

chessbyte commented:
@blomquisg What is the status of this PR?

chessbyte commented:
@blomquisg What is the status of this PR?

kbrock commented:
@fryguy Does smartstate work with anything other than vmware?
It suggests it works with openstack, but I don't see anything that suggests it would work. (not `supports` or similar method calls)

Not sure if moving this code out of here and into vmware repo makes sense,
or if moving this code into a smartstate module (or repo) would make sense.

Or if this would help us in the short term and is even a priority.

Fryguy commented:
According to the [Capabilities Matrix](https://www.manageiq.org/docs/reference/latest/capabilities_matrix/): vmware, rhv/ovirt, openstack, aws and azure.
